fix simd128.h build error on ARM64 Windows (#984)

Changqing-JING opened the pull request:
fix #984

mr-c commented:
Thank you for this, @Changqing-JING ; do you know of a free CI service we can use to test this with the Microsoft compiler under Arm Windows?

Changqing-JING commented:
@mr-c Thank you for review.

I found another bug in function `simde_wasm_i8x16_shl'. I have pushed the fix in new commit.

## About CI
I'm sorry I don't have a free Widonws ARM64 CI server either.

## About Test on Window Arm64
I still didn't run full test on Arm64 yet. Because as far as I know, both meson and ninja didn't provide pre-compiled installer for Windows ARM64. I'm not sure if they support build from source on Windows ARM64. 
I can try it when I have time.

Changqing-JING commented:
I'm currently working on a Wabt fix. This repo refers simde as submodule and its tests failed on ARM64 windows. 
https://github.com/WebAssembly/wabt/pull/2140

Since the wabt is a WebAssemlby tooling, so my fix are focus on the wasm/simd128.h. In current commit, all bugs blocking wabt are fixed.

Could you help to merge this PR first? And we can evaluate solutions for other parts later.

Changqing-JING commented:
@mr-c 
I'm currently working on a Wabt fix. This repo refers simde as submodule and its tests failed on ARM64 windows. 
https://github.com/WebAssembly/wabt/pull/2140

Since the wabt is a WebAssemlby tooling, so my fix are focus on the wasm/simd128.h. In current commit, all bugs blocking wabt are fixed.

Could you help to merge this PR first? And we can evaluate solutions for other parts later.

mr-c commented:
Thanks for letting me know. I'll do some extra arm64 testing later this weekend. If that passes , I'll merge.

Changqing-JING commented:
I have made further tests on Windows 11 ARM64
## Environment setup
Luckily both ninja and meson works on ARM64 windows.
### Install Ninja on ARM64 windows
```shell
git clone https://github.com/ninja-build/ninja.git
cd ninja
mkdir build
cd build
cmake -G "Visual Studio 17 2022" -A ARM64 -Bbuild-cmake ..
cmake --build build-cmake
```
Add ${workspace_abs_path}\build\build-cmake\Debug to PATH

### Install meson on ARM64 windows
Meson is a CPU independent python package
```
python -m pip install meson
```

## Test steps
"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvars64.bat" arm64
mkdir build
cd build
meson setup ..
ninja test -k 100000 > simde_test_failed_on_windows_arm64.txt
```
Note: "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvars64.bat" shows error message, 
```
[ERROR:VsDevCmd.bat] *** VsDevCmd.bat encountered errors. Environment may be incomplete and/or incorrect. ***
[ERROR:VsDevCmd.bat] In an uninitialized command prompt, please 'set VSCMD_DEBUG=[value]' and then re-run
[ERROR:VsDevCmd.bat] vsdevcmd.bat [args] for additional details.
[ERROR:VsDevCmd.bat] Where [value] is:
[ERROR:VsDevCmd.bat]    1 : basic debug logging
[ERROR:VsDevCmd.bat]    2 : detailed debug logging
[ERROR:VsDevCmd.bat]    3 : trace level logging. Redirection of output to a file when using this level is recommended.
[ERROR:VsDevCmd.bat] Example: set VSCMD_DEBUG=3
[ERROR:VsDevCmd.bat]          vsdevcmd.bat > vsdevcmd.trace.txt 2>&1
```
but it doesn't matter, just ignore it.

### Test result
The ninja build failed with many errors. For example
```
[1150/3069] Compiling C object test/arm/neon/cge-native-c.exe.p/cge.c.obj
FAILED: test/arm/neon/cge-native-c.exe.p/cge.c.obj 
"cl" "-Itest\arm\neon\cge-native-c.exe.p" "-Itest\arm\neon" "-I..\test\arm\neon" "-I." "-I.." "/MDd" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "/openmp" "-DSIMDE_CONSTRAINED_COMPILATION" "-DSIMDE_TEST_BARE" "/Fdtest\arm\neon\cge-native-c.exe.p\cge.c.pdb" /Fotest/arm/neon/cge-native-c.exe.p/cge.c.obj "/c" ../test/arm/neon/cge.c
C:\workspace\simde\simde\arm\neon/ext.h(42): warning C4003: not enough arguments for function-like macro invocation 'vext_f32'
C:\workspace\simde\simde\arm\neon/ext.h(42): warning C4002: too many arguments for function-like macro invocation '__float32x2_t_to_n64'
C:\workspace\simde\simde\arm\neon/ext.h(42): error C2440: 'function': cannot convert from 'int' to '__n64'
C:\workspace\simde\simde\arm\neon/ext.h(42): warning C4024: 'neon_ext32': different types for formal and actual parameter 2
C:\workspace\simde\simde\arm\neon/ext.h(42): error C2059: syntax error: ')'
C:\workspace\simde\simde\arm\neon/ext.h(113): warning C4003: not enough arguments for function-like macro invocation 'vext_s8'
C:\workspace\simde\simde\arm\neon/ext.h(113): warning C4002: too many arguments for function-like macro invocation 
```
Full log here:
[simde_test_failed_on_windows_arm64.txt](https://github.com/simd-everywhere/simde/files/10610793/simde_test_failed_on_windows_arm64.txt)

### Root cause analyze:
#include <arm_neon.h> in Windows and Linux are different. In linux, vext_f32 is a function, but in Windows, vext_f32 is a macro.

This bug can be reproduced with follow minimal reproduce example:
Save follow code as main.c
```C
#include <arm_neon.h>

#define SIMDE_CONSTIFY_2_(func_name, result, default_case, imm, ...) \
  do { \
    switch(imm) { \
      case 0: result = func_name(__VA_ARGS__, 0); break; \
      case 1: result = func_name(__VA_ARGS__, 1); break; \
      default: result = default_case; break; \
    } \
  } while (0)


float32x2_t function1(float32x2_t x, float32x2_t y, int z){
  float32x2_t r;
  return r;
}


int main(){
  float32x2_t a;
  float32x2_t b;
  int n;
  float32x2_t res;
  float32x2_t errorValue;
	SIMDE_CONSTIFY_2_(function1, res, errorValue, n, a, b);
	SIMDE_CONSTIFY_2_(vext_f32, res, errorValue, n, a, b);

 return 0;
}
````
#### On linux, vext_f32 is a function, so the result is correct:
Build with GCC
```
gcc -E -o build/main.temp.linux.c main.c
```
The C code after pre process is:
```C
int main(){
  float32x2_t a;
  float32x2_t b;
  int n;
  float32x2_t res;
  float32x2_t errorValue;
 do { switch(n) { case 0: res = function1(a, b, 0); break; case 1: res = function1(a, b, 1); break; default: res = errorValue; break; } } while (0);
 do { switch(n) { case 0: res = vext_f32(a, b, 0); break; case 1: res = vext_f32(a, b, 1); break; default: res = errorValue; break; } } while (0);

 return 0;
}
```
Full code here:
[main.temp.linux.c.txt](https://github.com/simd-everywhere/simde/files/10610819/main.temp.linux.c.txt)

#### On windows, vext_f32 is a macro, which leads to incorrect result:
Build with MSVC
```batch
"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvars64.bat" arm64
cl /E main.c > build/main.temp.windows.c
```
The C code after pre process is:
```C
int main(){
  float32x2_t a;
  float32x2_t b;
  int n;
  float32x2_t res;
  float32x2_t errorValue;
	do { switch(n) { case 0: res = function1(a, b, 0); break; case 1: res = function1(a, b, 1); break; default: res = errorValue; break; } } while (0);
	do { switch(n) { case 0: res = (neon_ext32((a), (0), )); break; case 1: res = (neon_ext32((a), (1), )); break; default: res = errorValue; break; } } while (0);

 return 0;
}
```
Full code here:
[main.temp.windows.c.txt](https://github.com/simd-everywhere/simde/files/10610838/main.temp.windows.c.txt)

Because in Windows, `vext_f32` is a macro define as this:
```C
#define vext_f32(Dn, Dm, pos)                    ( __static_assert((pos) >= 0 && (pos) < 2, "invalid position value"), __neon_DdDnDm( 0xf2b00000 | _NENC_11_8((pos) * 4), (Dn), (Dm)) )
```
Full code of arm_neo.h on Windows:
[arm_neon.h.txt](https://github.com/simd-everywhere/simde/files/10610904/arm_neon.h.txt)

To solve this problem, my proposal is to create a function vext_f32_wrapper, then call this function with SIMDE_CONSTIFY_2_ and call vext_f32 in this wrapper. What do you think about?

Changqing-JING commented:
@mr-c 
Good news, All tests on ARM64 Windows. Full logs here:
[simde_windows_arm64_test_log.txt](https://github.com/simd-everywhere/simde/files/10671379/simde_windows_arm64_test_log.txt)
[simde_windows_arm64_build_log.txt](https://github.com/simd-everywhere/simde/files/10671392/simde_windows_arm64_build_log.txt)


The reason of previous compile failure is: MSVC compile VA_ARGS in legacy mode by default. The legacy mode handles VA_ARGS different with C11, so we need to set environments to let MSVC to use C++.
```
set CFLAGS="/std:c11"
set CXXFLAGS="/Zc:preprocessor"
```
But there is a follow up problem that a bug of MSVC leads to some test failure. I have reported this bug to Microsoft. 
https://developercommunity.visualstudio.com/t/In-arm64_neonh-vsqaddb_u8-vsqaddh_u16/10271747

According to my experience, Microsoft won't release a fix of Windows SDK/compiler in short time. So as a workaround, I redefine the macro in simde code. We can remove the workaround after Microsoft release the fix.

Since currently all tests are passed, could you merge this PR?

mr-c commented:
Hello again @Changqing-JING  ; thank you for your persistence! I'm seeing build errors on aarch64 locally; I'll add a CI job to cross-build, so you can see them as well

<details>
<summary>big log excerpt</summary>
<pre>
[611/701] Compiling C object test/mips/msa/add_a-native-c.p/add_a.c.o
FAILED: test/mips/msa/add_a-native-c.p/add_a.c.o 
aarch64-linux-gnu-gcc-10 -Itest/mips/msa/add_a-native-c.p -Itest/mips/msa -I../../../usr/local/src/simde/test/mips/msa -I. -I../../../usr/local/src/simde -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c99 -g -march=armv8-a+simd+crypto+crc -Wextra -Werror -fopenmp-simd -DSIMDE_CONSTRAINED_COMPILATION -DSIMDE_ENABLE_OPENMP -Wno-psabi -DSIMDE_TEST_BARE -MD -MQ test/mips/msa/add_a-native-c.p/add_a.c.o -MF test/mips/msa/add_a-native-c.p/add_a.c.o.d -o test/mips/msa/add_a-native-c.p/add_a.c.o -c ../../../usr/local/src/simde/test/mips/msa/add_a.c
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:33,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/adds.h: In function 'simde_msa_adds_u_b':
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/adds.h:285:7: note: use '-flax-vector-conversions' to permit conversions between vectors with differing element types or numbers of subparts
  285 |       r_.neon = vqaddq_u8(a_.neon, b_.neon);
      |       ^~
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/adds.h:285:29: error: incompatible type for argument 1 of 'vqaddq_u8'
  285 |       r_.neon = vqaddq_u8(a_.neon, b_.neon);
      |                           ~~^~~~~
      |                             |
      |                             int8x16_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:2768:23: note: expected 'uint8x16_t' but argument is of type 'int8x16_t'
 2768 | vqaddq_u8 (uint8x16_t __a, uint8x16_t __b)
      |            ~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:33,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/adds.h:285:38: error: incompatible type for argument 2 of 'vqaddq_u8'
  285 |       r_.neon = vqaddq_u8(a_.neon, b_.neon);
      |                                    ~~^~~~~
      |                                      |
      |                                      int8x16_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:2768:39: note: expected 'uint8x16_t' but argument is of type 'int8x16_t'
 2768 | vqaddq_u8 (uint8x16_t __a, uint8x16_t __b)
      |                            ~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:33,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/adds.h: In function 'simde_msa_adds_u_h':
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/adds.h:322:30: error: incompatible type for argument 1 of 'vqaddq_u16'
  322 |       r_.neon = vqaddq_u16(a_.neon, b_.neon);
      |                            ~~^~~~~
      |                              |
      |                              int16x8_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:2775:24: note: expected 'uint16x8_t' but argument is of type 'int16x8_t'
 2775 | vqaddq_u16 (uint16x8_t __a, uint16x8_t __b)
      |             ~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:33,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/adds.h:322:39: error: incompatible type for argument 2 of 'vqaddq_u16'
  322 |       r_.neon = vqaddq_u16(a_.neon, b_.neon);
      |                                     ~~^~~~~
      |                                       |
      |                                       int16x8_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:2775:40: note: expected 'uint16x8_t' but argument is of type 'int16x8_t'
 2775 | vqaddq_u16 (uint16x8_t __a, uint16x8_t __b)
      |                             ~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:33,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/adds.h: In function 'simde_msa_adds_u_w':
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/adds.h:376:30: error: incompatible type for argument 1 of 'vqaddq_u32'
  376 |       r_.neon = vqaddq_u32(a_.neon, b_.neon);
      |                            ~~^~~~~
      |                              |
      |                              int32x4_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:2782:24: note: expected 'uint32x4_t' but argument is of type 'int32x4_t'
 2782 | vqaddq_u32 (uint32x4_t __a, uint32x4_t __b)
      |             ~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:33,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/adds.h:376:39: error: incompatible type for argument 2 of 'vqaddq_u32'
  376 |       r_.neon = vqaddq_u32(a_.neon, b_.neon);
      |                                     ~~^~~~~
      |                                       |
      |                                       int32x4_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:2782:40: note: expected 'uint32x4_t' but argument is of type 'int32x4_t'
 2782 | vqaddq_u32 (uint32x4_t __a, uint32x4_t __b)
      |                             ~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:33,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/adds.h: In function 'simde_msa_adds_u_d':
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/adds.h:407:30: error: incompatible type for argument 1 of 'vqaddq_u64'
  407 |       r_.neon = vqaddq_u64(a_.neon, b_.neon);
      |                            ~~^~~~~
      |                              |
      |                              int64x2_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:2789:24: note: expected 'uint64x2_t' but argument is of type 'int64x2_t'
 2789 | vqaddq_u64 (uint64x2_t __a, uint64x2_t __b)
      |             ~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:33,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/adds.h:407:39: error: incompatible type for argument 2 of 'vqaddq_u64'
  407 |       r_.neon = vqaddq_u64(a_.neon, b_.neon);
      |                                     ~~^~~~~
      |                                       |
      |                                       int64x2_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:2789:40: note: expected 'uint64x2_t' but argument is of type 'int64x2_t'
 2789 | vqaddq_u64 (uint64x2_t __a, uint64x2_t __b)
      |                             ~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:37,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/and.h: In function 'simde_msa_and_v':
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/and.h:52:28: error: incompatible type for argument 1 of 'vandq_u8'
   52 |       r_.neon = vandq_u8(a_.neon, b_.neon);
      |                          ~~^~~~~
      |                            |
      |                            int8x16_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:1612:22: note: expected 'uint8x16_t' but argument is of type 'int8x16_t'
 1612 | vandq_u8 (uint8x16_t __a, uint8x16_t __b)
      |           ~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:37,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/and.h:52:37: error: incompatible type for argument 2 of 'vandq_u8'
   52 |       r_.neon = vandq_u8(a_.neon, b_.neon);
      |                                   ~~^~~~~
      |                                     |
      |                                     int8x16_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:1612:38: note: expected 'uint8x16_t' but argument is of type 'int8x16_t'
 1612 | vandq_u8 (uint8x16_t __a, uint8x16_t __b)
      |                           ~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:38,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/andi.h: In function 'simde_msa_andi_b':
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/andi.h:51:28: error: incompatible type for argument 1 of 'vandq_u8'
   51 |       r_.neon = vandq_u8(a_.neon, vdupq_n_u8(HEDLEY_STATIC_CAST(uint8_t, imm0_255)));
      |                          ~~^~~~~
      |                            |
      |                            int8x16_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:1612:22: note: expected 'uint8x16_t' but argument is of type 'int8x16_t'
 1612 | vandq_u8 (uint8x16_t __a, uint8x16_t __b)
      |           ~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:40,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/madd.h: In function 'simde_msa_fmadd_w':
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/madd.h:53:29: error: incompatible type for argument 1 of 'vfmaq_f32'
   53 |       r_.neon = vfmaq_f32(a_.neon, c_.neon, b_.neon);
      |                           ~~^~~~~
      |                             |
      |                             int32x4_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:16670:24: note: expected 'float32x4_t' but argument is of type 'int32x4_t'
16670 | vfmaq_f32 (float32x4_t __a, float32x4_t __b, float32x4_t __c)
      |            ~~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:40,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/madd.h:53:38: error: incompatible type for argument 2 of 'vfmaq_f32'
   53 |       r_.neon = vfmaq_f32(a_.neon, c_.neon, b_.neon);
      |                                    ~~^~~~~
      |                                      |
      |                                      int32x4_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:16670:41: note: expected 'float32x4_t' but argument is of type 'int32x4_t'
16670 | vfmaq_f32 (float32x4_t __a, float32x4_t __b, float32x4_t __c)
      |                             ~~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:40,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/madd.h:53:47: error: incompatible type for argument 3 of 'vfmaq_f32'
   53 |       r_.neon = vfmaq_f32(a_.neon, c_.neon, b_.neon);
      |                                             ~~^~~~~
      |                                               |
      |                                               int32x4_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:16670:58: note: expected 'float32x4_t' but argument is of type 'int32x4_t'
16670 | vfmaq_f32 (float32x4_t __a, float32x4_t __b, float32x4_t __c)
      |                                              ~~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:40,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/madd.h: In function 'simde_msa_fmadd_d':
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/madd.h:98:29: error: incompatible type for argument 1 of 'vfmaq_f64'
   98 |       r_.neon = vfmaq_f64(a_.neon, c_.neon, b_.neon);
      |                           ~~^~~~~
      |                             |
      |                             int64x2_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:16677:24: note: expected 'float64x2_t' but argument is of type 'int64x2_t'
16677 | vfmaq_f64 (float64x2_t __a, float64x2_t __b, float64x2_t __c)
      |            ~~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:40,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/madd.h:98:38: error: incompatible type for argument 2 of 'vfmaq_f64'
   98 |       r_.neon = vfmaq_f64(a_.neon, c_.neon, b_.neon);
      |                                    ~~^~~~~
      |                                      |
      |                                      int64x2_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:16677:41: note: expected 'float64x2_t' but argument is of type 'int64x2_t'
16677 | vfmaq_f64 (float64x2_t __a, float64x2_t __b, float64x2_t __c)
      |                             ~~~~~~~~~~~~^~~
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:40,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/madd.h:98:47: error: incompatible type for argument 3 of 'vfmaq_f64'
   98 |       r_.neon = vfmaq_f64(a_.neon, c_.neon, b_.neon);
      |                                             ~~^~~~~
      |                                               |
      |                                               int64x2_t
In file included from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-features.h:346,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/../../simde-common.h:43,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa/types.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/../../../simde/mips/msa.h:30,
                 from ../../../usr/local/src/simde/test/mips/msa/test-msa.h:4,
                 from ../../../usr/local/src/simde/test/mips/msa/add_a.c:3:
/usr/lib/gcc-cross/aarch64-linux-gnu/10/include/arm_neon.h:16677:58: note: expected 'float64x2_t' but argument is of type 'int64x2_t'
16677 | vfmaq_f64 (float64x2_t __a, float64x2_t __b, float64x2_t __c)
      |                                              ~~~~~~~~~~~~^~~
</pre>
</details>

mr-c commented:
Looks like the ARM64 Windows issues were already encountered by @nemequ https://github.com/simd-everywhere/simde/commit/3d8a516d7e09c9e91f48f2fda304a09a5770c785

mr-c commented:
@Changqing-JING Do you have ideas how to fix the errors at https://github.com/simd-everywhere/simde/actions/runs/4114050091/jobs/7100937430#step:5:1332 ?

Changqing-JING commented:
@mr-c The Build error should be the preprocessor error which I mentioned yesterday. I have added a fix, let's see if it works.
I also fixed the arm64 linux build error in the commit, now linux arm64 tests all passed
[linux_arm_test_success.txt](https://github.com/simd-everywhere/simde/files/10681066/linux_arm_test_success.txt)

Changqing-JING commented:
It seems the CFLAGS in environment is not used if a cross file is set during meson setup. I have changed the /std:c11 to the cross file, now the x64_arm64 cross compile can success on my local x64 laptop.
```
Ok:                 0
Expected Fail:      0
Fail:               0
Unexpected Pass:    0
Skipped:            1534
Timeout:            0
```

Changqing-JING commented:
@mr-c Does the CI look good now?

mr-c commented on simde/arm/neon/sqadd.h:
```diff
@@ -30,6 +30,19 @@
 #include "types.h"
 #include <limits.h>
 
+// Workaround on ARM64 windows due to windows SDK bug
+// https://developercommunity.visualstudio.com/t/In-arm64_neonh-vsqaddb_u8-vsqaddh_u16/10271747?sort=newest
+#if (defined _MSC_VER) && (defined SIMDE_ARM_NEON_A64V8_NATIVE)
+#undef vsqaddb_u8
+#define vsqaddb_u8(src1, src2) neon_usqadds8(__uint8ToN8_v(src1), __int8ToN8_v(src2)).n8_u8[0]
+#undef vsqaddh_u16
+#define vsqaddh_u16(src1, src2) neon_usqadds16(__uint16ToN16_v(src1), __int16ToN16_v(src2)).n16_u16[0]
+#undef vsqadds_u32
+#define vsqadds_u32(src1, src2) _CopyUInt32FromFloat(neon_usqadds32(_CopyFloatFromUInt32(src1), _CopyFloatFromInt32(src2)))
+#undef vsqaddd_u64
+#define vsqaddd_u64(src1, src2) neon_usqadds64(__uint64ToN64_v(src1), __int64ToN64_v(src2)).n64_u64[0]
+#endif
```
I would prefer that these were apart of the relevant `simde_vsqadd*` functions below

mr-c commented on simde/arm/neon/uqadd.h:
```diff
@@ -33,6 +33,17 @@ HEDLEY_DIAGNOSTIC_PUSH
 SIMDE_DISABLE_UNWANTED_DIAGNOSTICS
 SIMDE_BEGIN_DECLS_
 
+// Workaround on ARM64 windows due to windows SDK bug
+// https://developercommunity.visualstudio.com/t/In-arm64_neonh-vsqaddb_u8-vsqaddh_u16/10271747?sort=newest
+#if (defined _MSC_VER) && (defined SIMDE_ARM_NEON_A64V8_NATIVE)
+#undef vuqaddh_s16
+#define vuqaddh_s16(src1, src2) neon_suqadds16(__int16ToN16_v(src1), __uint16ToN16_v(src2)).n16_i16[0]
+#undef vuqadds_s32
+#define vuqadds_s32(src1, src2) _CopyInt32FromFloat(neon_suqadds32(_CopyFloatFromInt32(src1), _CopyFloatFromUInt32(src2)))
+#undef vuqaddd_s64
+#define vuqaddd_s64(src1, src2) neon_suqadds64(__int64ToN64_v(src1), __uint64ToN64_v(src2)).n64_i64[0]
+#endif
+
```
Like sqadd, I would also like these redefines to be associated with the relevant `simde_vuqadd*` functions below.

mr-c commented on simde/mips/msa/types.h:
```diff
@@ -49,7 +49,7 @@ typedef union {
     __m128i m128i;
   #endif
-  #if defined(SIMDE_MIPS_MSA_A32V7_NATIVE)
+  #if defined(SIMDE_ARM_NEON_A32V7_NATIVE)
```
Great catch, thank you!

mr-c commented on simde/mips/msa/add_a.h:
```diff
@@ -38,8 +38,6 @@ simde_v16i8
 simde_msa_add_a_b(simde_v16i8 a, simde_v16i8 b) {
   #if defined(SIMDE_MIPS_MSA_NATIVE)
     return __msa_add_a_b(a, b);
-  #elif defined(SIMDE_ARM_NEON_A32V7_NATIVE)
-    return vaddq_s8(vabsq_s8(a), vabsq_s8(b));
```
If `SIMDE_MIPS_MSA_A32V7_NATIVE` is renamed to `SIMDE_ARM_NEON_A32V7_NATIVE` on line 233 of `simde/mips/msa/types.h` then I don't think these changes are needed as `simde_v16i8` will be a neon `int8x16_t` and not something that needs converting to a  `simde_v16i8_private` to access the `.neon` attribute.

Changqing-JING replied:
The reason why I changed it to r_.neon = xxx(a_.neon, b_neon):
The key point is not the a, and b. The problem is r_.
Return type of vaddq_s8 can't be casted to simde_v16i8 implicitly on MSVC build. 
I used `r_.neon = xxx(a_.neon, b_.neon);` because the definition of r_ can be reused.

Otherwise, we need to write:
```C
#elif defined(SIMDE_ARM_NEON_A32V7_NATIVE)
    simde_v16i8 r_;
    r.neon = vaddq_s8(vabsq_s8(a), vabsq_s8(b));
    return r_;
```

Or is there other better implementation?

mr-c replied:
Using this function as an example, its return type is `simde_v16i8` ; as I said above, that is typedef to be a neon `int8x16_t` if my recommended change (`SIMDE_MIPS_MSA_A32V7_NATIVE` is changed to `SIMDE_ARM_NEON_A32V7_NATIVE`) is made. No cast problems.

https://github.com/simd-everywhere/simde/blob/f0f3d09547fefaca1134923ce0c0af1ca4a176b3/simde/mips/msa/types.h#L233-L234

Changqing-JING replied:
Reverted the files back

mr-c commented on simde/wasm/simd128.h:
```diff
@@ -3579,7 +3579,7 @@ simde_wasm_i8x16_shl (simde_v128_t a, uint32_t count) {
 
     #if defined(SIMDE_ARM_NEON_A32V7_NATIVE)
-      r_.neon_i8 = vshlq_s8(a_.neon_i8, vdupq_n_s8(HEDLEY_STATIC_CAST(int8_t, count)));
+      r_.neon_i8 = vshlq_s8(a_.neon_i8, vdupq_n_s8(HEDLEY_STATIC_CAST(int8_t, count & 7)));
```
Are these changes needed for all arm64 platforms, or just MS Windows?

Changqing-JING replied:
Yes, all arm platforms, Wasm standard requires
![image](https://user-images.githubusercontent.com/59640930/217563129-19d3401e-a1e2-417e-bcff-16c3289431b0.png)

https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md

Changqing-JING replied:
I just found another problem:
We used `count & 7` at line 3582 and line 3590, but `count & 7` != `count % 7`. It passed the test case of WABT because WABT tests count = 8, then `count & 7` and `count % 7` happens equal to 0.

We need to change all  `& 7` to `%7` in this file.

Changqing-JING replied:
Test cases added

mr-c commented:
> @mr-c Does the CI look good now?

CI looks great, thank you! I left some comments about the code itself

mr-c commented:
```
1416/1534 wasm_simd128/shl/native/c               ERROR           0.03s   exit status 1
>>> MALLOC_PERTURB_=243 MESON_EXE_WRAPPER=/usr/bin/v8 /usr/bin/v8 /home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
stderr:
exiting due to exception: RuntimeError: unreachable,RuntimeError: unreachable
    at wasm://wasm/0a7ac7f6:wasm-function[5]:0x64c
    at wasm://wasm/0a7ac7f6:wasm-function[8]:0x949
    at /home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:776:22
    at callMain (/home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:1552:15)
    at doRun (/home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:1605:23)
    at run (/home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:1620:5)
    at runCaller (/home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:1537:19)
    at removeRunDependency (/home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:685:7)
    at receiveInstance (/home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:899:5)
    at receiveInstantiationResult (/home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:917:5)

1424/1534 wasm_simd128/shr/native/c               ERROR           0.03s   exit status 1
>>> MESON_EXE_WRAPPER=/usr/bin/v8 MALLOC_PERTURB_=35 /usr/bin/v8 /home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
stderr:
exiting due to exception: RuntimeError: unreachable,RuntimeError: unreachable
    at wasm://wasm/00011bd2:wasm-function[9]:0xa40
    at wasm://wasm/00011bd2:wasm-function[5]:0x3a4
    at wasm://wasm/00011bd2:wasm-function[15]:0x1033
    at /home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js:776:22
    at callMain (/home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js:1552:15)
    at doRun (/home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js:1605:23)
    at run (/home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js:1620:5)
    at runCaller (/home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js:1537:19)
    at removeRunDependency (/home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js:685:7)
    at receiveInstance (/home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js:899:5)
```
https://github.com/simd-everywhere/simde/actions/runs/4132069468/jobs/7140340227#step:9:1474

mr-c commented on simde/wasm/simd128.h:
```diff
@@ -3579,9 +3579,9 @@ simde_wasm_i8x16_shl (simde_v128_t a, uint32_t count) {
 
     #if defined(SIMDE_ARM_NEON_A32V7_NATIVE)
-      r_.neon_i8 = vshlq_s8(a_.neon_i8, vdupq_n_s8(HEDLEY_STATIC_CAST(int8_t, count)));
+      r_.neon_i8 = vshlq_s8(a_.neon_i8, vdupq_n_s8(HEDLEY_STATIC_CAST(int8_t, count & 7)));
```
```suggestion
      r_.neon_i8 = vshlq_s8(a_.neon_i8, vdupq_n_s8(HEDLEY_STATIC_CAST(int8_t, count % 7)));
```
Shouldn't it be the modulo operator and not the bitwise AND operator?

Changqing-JING replied:
Number 2 power x are special, n % 8 == n & 7. Usually compiler will optimize code like this because bit and is much faster than div.
You can see in this example, gcc compile `argc % 8` to `argc & 7`
https://godbolt.org/z/GP69KW9YG

mr-c replied:
Got it, thanks! Sorry for my lack of C experience, I'm the backup maintainer for SIMDe

Changqing-JING commented:
> ```
> 1416/1534 wasm_simd128/shl/native/c               ERROR           0.03s   exit status 1
> >>> MALLOC_PERTURB_=243 MESON_EXE_WRAPPER=/usr/bin/v8 /usr/bin/v8 /home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js
> ――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
> stderr:
> exiting due to exception: RuntimeError: unreachable,RuntimeError: unreachable
>     at wasm://wasm/0a7ac7f6:wasm-function[5]:0x64c
>     at wasm://wasm/0a7ac7f6:wasm-function[8]:0x949
>     at /home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:776:22
>     at callMain (/home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:1552:15)
>     at doRun (/home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:1605:23)
>     at run (/home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:1620:5)
>     at runCaller (/home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:1537:19)
>     at removeRunDependency (/home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:685:7)
>     at receiveInstance (/home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:899:5)
>     at receiveInstantiationResult (/home/runner/work/simde/simde/build/test/wasm/simd128/shl-native-c.js:917:5)
> 
> 1424/1534 wasm_simd128/shr/native/c               ERROR           0.03s   exit status 1
> >>> MESON_EXE_WRAPPER=/usr/bin/v8 MALLOC_PERTURB_=35 /usr/bin/v8 /home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js
> ――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
> stderr:
> exiting due to exception: RuntimeError: unreachable,RuntimeError: unreachable
>     at wasm://wasm/00011bd2:wasm-function[9]:0xa40
>     at wasm://wasm/00011bd2:wasm-function[5]:0x3a4
>     at wasm://wasm/00011bd2:wasm-function[15]:0x1033
>     at /home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js:776:22
>     at callMain (/home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js:1552:15)
>     at doRun (/home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js:1605:23)
>     at run (/home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js:1620:5)
>     at runCaller (/home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js:1537:19)
>     at removeRunDependency (/home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js:685:7)
>     at receiveInstance (/home/runner/work/simde/simde/build/test/wasm/simd128/shr-native-c.js:899:5)
> ```
> 
> https://github.com/simd-everywhere/simde/actions/runs/4132069468/jobs/7140340227#step:9:1474

This bug is inside emcc part, it should not be caused by today's update, but today's update find a hidden bug.
I have looked into it for 1 hour but didn't find the root cause.
Could you also have a look in mean while?

mr-c commented:
> This bug is inside emcc part, it should not be caused by today's update, but today's update find a hidden bug. I have looked into it for 1 hour but didn't find the root cause. Could you also have a look in mean while?

If you think it is a recent bug, then I can change the CI setup to use a different release, just tell me the version to pass to emsdk.

Changqing-JING commented:
@mr-c I have confirmed that the test failure is caused by a bug in clang. I have reported it to llvm-project:
https://github.com/llvm/llvm-project/issues/60655

This bug exists in clang for long time, and still exists in main HEAD.

I will add our local work around to fix the test

mr-c commented:
> @mr-c I have confirmed that the test failure is caused by a bug in clang. I have reported it to llvm-project: [llvm/llvm-project#60655](https://github.com/llvm/llvm-project/issues/60655)
> 
> This bug exists in clang for long time, and still exists in main HEAD.

Thanks for tracking that down and filing the bug report!

> I have added a local work around to fix the test

If `-O2` is the problem, then I'm fine to change `docker/cross-files/emscripten.cross` to use `-O1` instead

https://github.com/simd-everywhere/simde/blob/73dddb724081ddd1315e94e3a8789df897154e3e/.github/workflows/ci.yml#L153

https://github.com/simd-everywhere/simde/blob/73dddb724081ddd1315e94e3a8789df897154e3e/docker/cross-files/emscripten.cross#L11

Changqing-JING commented:
Add a count & 63 patch is better than downgrade optimization level in CI. Because downgrade optimization level in CI config can just let CI pass, our users may still use -O2 to build their product. Then the their product may crash. Patch it on simde side is more user friendly.

mr-c commented:
> Add a `count & 63` patch is better than downgrade optimization level in CI. Because downgrade optimization level in CI config can just let CI pass, our users may still use -O2 to build their product. Then the their product may crash. Patch it on simde side is more user friendly.

Sure, but then lets make it Clang specific like we work around other clang bugs

Changqing-JING commented:
Do you mean add a `#ifdef __clang__`?
Currently Clang is our only supported compiler for C to wasm. Only clang target wasm32 can pass macro `SIMDE_WASM_SIMD128_NATIVE`. Do we still need to add `#ifdef __clang__`?

mr-c commented:
> Do you mean add a `#ifdef __clang__`?
> Currently Clang is our only supported compiler for C to wasm. Only clang target wasm32 can pass macro `SIMDE_WASM_SIMD128_NATIVE`. Do we still need to add `#ifdef __clang__`?

Whoops, you are correct. I see the emscripten build has passed!

Changqing-JING commented:
The sleef test failed due to today it's running on a different CPU.
Today:
```
model name	: Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
```
Yesterday:
Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz

The test failure should be caused by CPU doesn't support some instructions.

Is simde project allow to merge this PR ignoring this test failure?

mr-c commented:
> The sleef test failed due to today it's running on a different CPU. Today:
> 
> ```
> model name	: Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
> ```
> 
> Yesterday: Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz
> 
> The test failure should be caused by CPU doesn't support some instructions.

Very interesting. The sleef job uses `-march=native`. Maybe the job got migrated to another CPU halfway through? I hit the 'retry' button.

I'll fix this, don't worry

mr-c reviewed (approved):
Thank you @Changqing-JING ! I will merge this with some small changes today or tomorrow

mr-c commented:
This PR has been merged as
- dad8cad30ad9a7f9e70e83de911b53fc6af020eb
- 0f988c9e519f51079e9f04a51b70e1ec7e6f5a5c
- b54dfcb139a09a800af31b60a3aba420468b0692
- 782d8164b31f03428824cea6c7eafe86c403b9bc
- 256d9dfcf0cc9e72a386e327f2c8cbe8087031bd
